openstack/datadog_checks/openstack/openstack.py

```diff
@@ -844,24 +856,53 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
             for i, avg in enumerate([1, 5, 15]):
                 self.gauge('openstack.nova.hypervisor_load.{0}'.format(avg), load_averages[i], tags=tags)
 
-    def get_all_server_ids(self, filter_by_host=None):
+    # Get all of the server IDs and their metadata and cache them
+    # After the first run, we will only get servers that have changed state since the last collection run
+    def get_all_servers(self, i_key, filter_by_host=None):
         query_params = {}
         if filter_by_host:
             query_params["host"] = filter_by_host
 
-        url = '{0}/servers'.format(self.get_nova_endpoint())
+        # If we don't have a timestamp for this instance, default to None
+        if i_key not in self.changes_since_time:
+            self._changes_since_time[i_key] = None
+
+        if self.changes_since_time[i_key]:
+            query_params['changes-since'] = changes_since_time
+
+        # Since we ultimately only perform actions on ACTIVE servers
+        # Lets filter to only those to reduce the payload we receive
+        active_query_params['status'] = 'ACTIVE,DELETED,SHUTOFF'
+
+        url = '{0}/servers/detail'.format(self.get_nova_endpoint())
         headers = {'X-Auth-Token': self.get_auth_token()}
 
         server_ids = []
         try:
+            # Get a list of active/deleted/shut-off servers
             resp = self._make_request_with_auth_fallback(url, headers, params=query_params)
-
-            server_ids = [s['id'] for s in resp['servers']]
+            self.changes_since_time[i_key] = datetime.utcnow().isoformat()
         except Exception as e:
             self.warning('Unable to get the list of all servers: {0}'.format(str(e)))
             raise e
 
-        return server_ids
+        for server in resp['servers']:
+            new_server = {}
+
+            new_server['server_id'] = server.get('id')
+            new_server['state'] = server.get('status')
+            new_server['server_name'] = server.get('name')
+            new_server['hypervisor_hostname'] = server.get('OS-EXT-SRV-ATTR:hypervisor_hostname')
+            new_server['tenant_id'] = server.get('tenant_id')
+            new_server['project_name'] = self.get_project_name_from_id(new_server['tenant_id'])
+
+            for server in self.servers[i_key]:
+                if server_id == server['id'] and state not in DIAGNOSTICABLE_STATES:
+                    self.servers[i_key].append(new_server)
+                elif server_id == server['id'] and state in REMOVED_STATES:
+                    self.servers[i_key][:] = [server for server in self.servers[i_key] if self.servers[i_key]('server_id') != server_id]
```

Comment:
As mentioned earlier, I'm not a fan of huge fan of using lists here, it doesn't seem like the right data structure as these lists can get quite large.